content/topics/java-specific/oop-basics/_index.md

```diff
@@ -1,15 +1,179 @@
+
+The following Java code shows how Encapsulation can be implemented:
+
+```
+public class Employee{
+
+    //private data member  
+    private String name;
+
+    //getter method for name  
+    public String getName(){
+        return name;
+    }Encapsulation is known as data-hiding.
+It is an OOP principle which states that all the attributes and behaviours (methods and functions) of an object should be grouped together in one datatype or class.
+This means that objects may be able to communicate with one another but are restricted to access some of the object's components directly.
+Publicly accessible methods are generally provided in the class (so-called [accessors and mutators.]
+    }
+
+    public static void main(String args[]) { 
+        //creating instance of the encapsulated class
+        Employee e=new Employee();
+        //setting value in the name member  
+        e.setName("Mbali");
+        //getting value of the name member  
+        System.out.println(e.getName());
+
+    }
+}  
+```
+[Abstraction vs Encapsulation](https://1.bp.blogspot.com/-ECYNAUTGGMk/WPQeY4EpFtI/AAAAAAAAIX8/j-Ji8N_mDz8-d72SasgNPnQD-nIlw-kiACLcB/s1600/Abstraction%2Bvs%2BEncapsulation%2B2.jpg)
+
+### Inheritance
+This OOP principle allows one class to inherit all the attributes and behaviors of another class.
+Inheritance has a hierarchy  relationship between superclasses (base classes )and subclasses
+
+The following Java code shows how Inheritance can be implemented:
+```
+//superclass
+public class Vehicle {
+    void printType(){
+        System.out.println("I am a Vehicle");
+    }
+}
+
+public class Car extends  Vehicle {
+    //Override method 
+    @Override
+    void printType() {
+        //call method in super class
+        super.printType();
+        System.out.println("I am a car");
+    }
+
+
+    public static void main(String[] args) {
+        //Create a car object
+        Car car = new Car();
+        //call method
+        car.printType();
+    }
+
+}
+
+```
+
+### Polymorphism
+Polymorphism in Java is a concept by which we can perform a single action in different ways. Polymorphism is derived from 2 Greek words: poly and morphs. The word "poly" means many and "morphs" means forms. So polymorphism means many forms.
+
+There are two types of polymorphism in Java: compile-time polymorphism and run-time polymorphism. We can perform polymorphism in java by method overloading and method overriding.
+#### Method overriding
+
+Overriding a method is when a method in the subclass has the same name and method signature as a method in the superclass. When overriding a method you are not allowed to make the method more private. 
+Example:
+You cannot make a public method protected.
+
+The following Java code shows how overriding can be implemented:
+```
+public class Animal {
```

Comment:
This code is really really messy. 

We need:
1. class to do one thing
2. classes to not be defined inside eas\ch other unless there is a damn good reason to do it (flat is better than nested)
3. main function to be in another class entirely (something like class MainProgram).  This is one way that "seperation of concerns" and "loose coupling" comes into play.

THings should only be together if they belong together. Otherwise they should be seperate.  Code is easier to reason about if it is seperated out into logical units.

Reply:
Better would be something like this:

Here is an Animal. Nice and simple to reason about
```
public class Animal {
    public void eat() {
        System.out.println("Generic animal is eating");
    }
}
```

Here is some reusable horse code
```
static class Horse extends Animal {
    //Override method
    @Override
    public void eat() {
        System.out.println("Horse is eating");
    }
    
}
```

Now the main program ties everything together
```
public class MainProgram {

    public static void main(String[] args) {
        //Create a car object
        Horse horse = new Horse();
        //call method
        horse.eat();
    }
}
```

Reply:
Applied